### ISSUE TITLE
Crash on Android 8

### ISSUE THREAD

**Issue by bershanskiy** [Tobi823/ffupdater#31]
FFUpdater 71.0.3 crashed on Android 8 with the following trace:
```
de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker$BackgroundNetworkException: fail to check FIREFOX_RELEASE due to cancelled job
	at de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker.doBackgroundCheck(BackgroundUpdateChecker.kt:47)
	at de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker$doBackgroundCheck$1.invokeSuspend(Unknown Source:11)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:32)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:113)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@deedef3
```

**Comment by Tobi823** [comment]
Does the error occur more frequently or was it just a one-time occurrence?
Or was the error notification text misleading/too complex/badly translated?
(You should have seen this (translated?) text:)
> It's likely that the background check failed due to a temporary network connectivity issue. But if this error occurs more than once a week, please report it. Click to view the error log.

I think that your error is a network connection error (and network connections are prone to errors). The app can either ignore these errors or show them to the user. I've decided that the best way is to show all errors to the users.
(Because I think that the developer of FairMail is right with this sentence from the FAQ: )
> FairEmail does not hide errors like similar apps often do, so it is easier to diagnose problems.

**Comment by gerryfrancis** [comment]
It happens with a Google Pixel 3a (Android 11), too:

de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker$BackgroundNetworkException: fail to check FIREFOX_RELEASE due to cancelled job
	at de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker.doBackgroundCheck(BackgroundUpdateChecker.kt:47)
	at de.marmaro.krt.ffupdater.notification.BackgroundUpdateChecker.doWork(BackgroundUpdateChecker.kt:21)
	at androidx.work.CoroutineWorker$startWork$1.invokeSuspend(CoroutineWorker.kt:69)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@e60dae4

**Comment by bershanskiy** [comment]
Sorry for the delay, I must have lost this issue in my notifications. Please feel free to @ me if I don't respond within a day.

> I think that your error is a network connection error (and network connections are prone to errors).

Yes, it is.

> Does the error occur more frequently or was it just a one-time occurrence?

It is spotty, when I filed this issue initially this error occured every time I restarted my phone, but then it went away. I troubleshooted it a bit and I think in my case the error was caused by race between OS services starting up, adblocker starting up (implemented as a local VPN interfce) and ffupdater looking for updates.

The worst-case scenario is this:
 - I restart the phone, possibly after replacing the battery
 - the following things happen in parallel:
   - the OS does not have any connection and is looking for cellular/Wi-Fi. Since I removed the battery prior, the OS does not know accurate time so has to query network for it (and treats all requests as "insecure" and therefore fails them).
   - the OS sends startup notification to ffupdater, which queries for updates
   - the OS sends startup notification to adblocker, which creates VPN interface, registers it, etc.

In this scenario ffupdater is almost guaranteed to initiate its request before the either two finish, causing request to fail.

> The app can either ignore these errors or show them to the user. I've decided that the best way is to show all errors to the users.

Ideally, I'd expect apps to avoid startup races. For example, notice that the device was just restarted, silence the first error, and retry *once* after a short delay. This way, system has time to initialize everything, find network, get the time, and let other apps establish VPN interfaces.
Although I'm not sure the extra complexity (easily 30+ lines of code) is worth the small UX improvement.

**Comment by Tobi823** [comment]
> Sorry for the delay, I must have lost this issue in my notifications.

No hard feelings, I was busy the last weeks :)

I don't know when exactly but I improved the app to retry up two times (the initial failure + two tries) when a network connection failed.

I'll further increase the number of retries to four (the initial failure > 5s pause > first retry > 5s pause > ... > fourth and last retry > show error notification).

------------------

And I "misused" the Android WorkManager library: 

WorkManager executes the (repeating) background update checks. This library re-schedules the background update check even after a smartphone reboot. 

However, I added code to manually schedule the background update check after a smartphone reboot. This caused the background update check to run immediately after the smartphone startup.

By removing this code, I'll reduce the likelihood of these startup races. (They might happen if the smartphone is off for more than six hours - except you changed the "check interval". But the increased number of retries and the "retry window" of 20 seconds should be enough.)

------------------

The problem should be fixed in the next version (maybe 72.1.0) - the "background network exception" notification should not appear after starting the smartphone.
